### DISCUSSION TITLE
Proposal: Reconsidering button variants

### DISCUSSION THREAD

**Discussion by muratcorlu** [Trendyol/baklava#216] [Ideas]
Recently, while reading [an article](https://www.smashingmagazine.com/2022/08/back-button-ux-design/), drawing below attracted my attention about how they style button variants:

<img width="781" alt="image" src="https://user-images.githubusercontent.com/127687/188449638-396fdea3-c163-47fe-96c9-38ed766d335e.png">

I noticed that how general style of the buttons here expresses their hierarchy and makes their `primary`, `secondary` and `tertiary` naming more meaningful, regardless of their colors. The example above uses our `contained` buttons as `primary`, our `outline` as `secondary` and our `text` as `tertiary`.

In contrast, the hierarchy of our current `primary` and `secondary` buttons is directly -and only- related to their colors. `outline` and `text` versions of the buttons doesn't mean a hierarchy in our buttons as opposed to their styles. 

<img width="393" alt="image" src="https://user-images.githubusercontent.com/127687/188450668-b3027d6d-ddde-4a17-855c-b9c2aec3c8da.png">

I feel that, this creates some confussion especially when we discuss over our color palette. A few months ago, we decided to remove `tertiary` button because difference between `secondary` and `tertiary` buttons was only the color and the -default- colors were very close to each other.

<img width="651" alt="image" src="https://user-images.githubusercontent.com/127687/188451518-6f7325d6-7fc8-4490-a17c-4fe1bcabe940.png">

But we also needed a button lower than `secondary` in the hierarchy in multiple cases.

Now I want to offer to re-considering our button variant naming. I feel like using `outline` as `secondary` variant and `text` as the `tertiary` variant makes more sense, easier to explain and easier to use.

<img width="590" alt="image" src="https://user-images.githubusercontent.com/127687/188452680-4128d3c2-44ee-408c-b00a-0df10117806d.png">

With those variants, it's very obvious to see the hierarchy between the buttons regardless of their colors.

Then maybe we can use colors as `kind` to give button to a thematic value. like `<bl-button variant="secondary" kind="success">Success</bl-button>` can be:

<img width="189" alt="image" src="https://user-images.githubusercontent.com/127687/188453194-48d40781-7415-43b2-aadd-58aeca5f2fee.png">

To conclude, I think there is a big value to use `primary`, `secondary` and `tertiary` variants as the replacement of current `contained`, `outline` and `text` **kind**s and using kind parameter to them with their colors like `primary` color = `default`, `secondary` color = `neutral`, `success` and `danger`. In this case we may also need to rename our `primary` and `secondary` colors.

| kind \ variant  | `primary` | `secondary` | `tertiary` |
| -------------- | --------- | ------------- | --------- |
| `default` | <img width="146" alt="image" src="https://user-images.githubusercontent.com/127687/188454809-601849c5-5d9a-421e-a68c-afd331663ae3.png"> | <img width="164" alt="image" src="https://user-images.githubusercontent.com/127687/188455371-41375fab-d129-4ee6-937c-01247df3f0d4.png"> | <img width="144" alt="image" src="https://user-images.githubusercontent.com/127687/188455455-330be0bf-3b26-40ad-b9a8-7db2cd24dce4.png"> |
| `neutral` | <img width="149" alt="image" src="https://user-images.githubusercontent.com/127687/188455648-502ae73c-a27a-4174-a1ff-f2aca09a984e.png"> | <img width="169" alt="image" src="https://user-images.githubusercontent.com/127687/188455722-8e6419fb-80fe-4d2f-9373-b77c0fe8b714.png"> | <img width="145" alt="image" src="https://user-images.githubusercontent.com/127687/188455774-5bb904ee-d83c-4ef1-819e-b42f784d7396.png"> |
| `success` | <img width="149" alt="image" src="https://user-images.githubusercontent.com/127687/188455890-5352fc60-18be-4138-b5c2-ecae1df33531.png"> | <img width="166" alt="image" src="https://user-images.githubusercontent.com/127687/188455926-cfc7e11d-a9fb-458b-ba9c-c8a48bd26785.png"> | <img width="139" alt="image" src="https://user-images.githubusercontent.com/127687/188455964-6e5c8f47-d0b0-4dbb-9cf2-f5d7c0858995.png"> |
| `danger` | <img width="148" alt="image" src="https://user-images.githubusercontent.com/127687/188456057-70d4f2f9-75de-48a9-9350-d669a9367da8.png"> | <img width="166" alt="image" src="https://user-images.githubusercontent.com/127687/188456100-e3820bbf-7797-4734-bfc2-9b6d663913dc.png"> | <img width="147" alt="image" src="https://user-images.githubusercontent.com/127687/188456164-12a78fd6-896d-42a7-a184-a2d3fbe58aa1.png"> |

What do you think?

@buseselvi @leventozen @DamlaDemir

**Comment by muratcorlu** [comment]
`kind` of the button, in this case, provides a "contextual theme" to the button without indicating any hierarchy. I'm not sure about the `default` and `neutral` naming, by the way.

**Comment by koulgar** [comment]
This seems like more practical than what we've currently. Honestly only thing that bothers me naming `kind` as a coloring/theme related attribute. Wouldn't it be clearer to name it something like `color`? 
Reason i'm suggesting it something like `color` or `theme` seems more applicable to use in other components to unify usage patterns. Does it make sense?

**Reply by muratcorlu** [reply]
I'm also not sure about the naming `kind` as the attribute as well, but `color` feels like it accepts a real color code instead of a color name in the palette. And `theme` is not convenient here since `success` or `danger` is not really a theme. We'll give theming options to Baklava users by setting their own design token variables. Then an attribute named `theme` will give the feel that you are setting a design theme for that component.

Maybe `context` can be another option?

**Comment by DamlaDemir** [comment]
This idea is better and clearer for variant usage. Actually, we discussed with @leventozen before about using the variant when trying to find a new prop name `kind`. We thought to give color or theme instead variant too. But it doesn't fit our purpose. I agree with you @muratcorlu  if we use the `color` it may mean we can give any color. When I glance at the other UI libraries I saw the names `intent`, `colorScheme` for color.

I think, maybe we can use `mode` instead of `kind`. 
`mode=default` `mode=dark` `mode=success` `mode=danger`

**Comment by leventozen** [comment]
Very good perspective. I agree with what you pointed out. However, I think we only need to replace `kind` with `variant` for the naming part and everything became ok to me.
Because I tend to think that `primary` is for primary actions. That is the kind we can say about the button.

Therefore, when we need to specify its `kind`, those are: `primary`, `secondary` and `tertiary`.
```js
<bl-button kind="secondary">secondary</bl-button>
```
It is a secondary action, means UX decided this action is `secondary`, ex: cancelling a dialog screen.
Additionally,
```js
<bl-button kind="primary" variant="danger">primary</bl-button>
```
It should make the user think that it is the primary action for this but with the danger variant, should think that it may cause some irreversible changes, let's say.

**Comment by buseselvi** [comment]
As the UX team we agree that it would be better to update the buttons as @muratcorlu  suggested, it would be more safe to determine the usage areas of the buttons according to the style rather than the color. 

Maybe we can call it `state`; default, neutral, success, and error terms seem to indicate a state.

**Comment by muratcorlu** [comment]
As always, naming things is one of the biggest challenges of our work 😊 Let me sum-up da suggestions for the button styles:

- `color`
- `theme`
- `mode`
- `context`
- `variant` (use `kind` for hierarchical names)
- `state`

Let's give a few more days for more ideas. In the meantime I'll have a look to more component libraries as examples and check dictionaries to understand the differences of the meaning of those words 😊

**Comment by muratcorlu** [comment]
Just for noting down here, we implemented those changes with #250
